protocols/request-response/src/cbor.rs

```diff
@@ -0,0 +1,150 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use crate::{Config, ProtocolSupport};
+use async_trait::async_trait;
+use futures::prelude::*;
+use futures::{AsyncRead, AsyncWrite};
+use libp2p_core::upgrade::{read_length_prefixed, write_length_prefixed};
+use libp2p_swarm::{NetworkBehaviour, StreamProtocol};
+use serde::{de::DeserializeOwned, Serialize};
+use std::{io, marker::PhantomData};
+
+#[derive(Debug, Clone)]
+pub struct Codec<Req, Resp> {
+    phantom: PhantomData<(Req, Resp)>,
+}
+
+const REQUEST_SIZE_MAXIMUM: usize = 1_000_000;
+const RESPONSE_SIZE_MAXIMUM: usize = 500_000_000;
+
+pub type OutEvent<Req, Resp> = crate::Event<Req, Resp>;
+
+#[derive(NetworkBehaviour)]
+#[behaviour(
+    to_swarm = "OutEvent<Req, Resp>",
+    prelude = "libp2p_swarm::derive_prelude"
+)]
+pub struct Behaviour<Req, Resp>
+where
+    Req: Send + Clone + Serialize + DeserializeOwned + 'static,
+    Resp: Send + Clone + Serialize + DeserializeOwned + 'static,
+{
+    inner: crate::Behaviour<Codec<Req, Resp>>,
+}
+
+#[async_trait]
+impl<Req, Resp> crate::Codec for Codec<Req, Resp>
+where
+    Req: Send + Clone + Serialize + DeserializeOwned,
+    Resp: Send + Clone + Serialize + DeserializeOwned,
+{
+    type Protocol = StreamProtocol;
+    type Request = Req;
+    type Response = Resp;
+
+    async fn read_request<T>(&mut self, _: &Self::Protocol, io: &mut T) -> io::Result<Req>
+    where
+        T: AsyncRead + Unpin + Send,
+    {
+        let vec = read_length_prefixed(io, REQUEST_SIZE_MAXIMUM).await?;
```

Comment:
@mxinden you mentioned in the past that you'd like to not prefix the length in req-res protocols. Do you see a way for how we can avoid this here?

Reply:
Why do we need it here?

1. The sender sends the request and closes the stream. (Note that `libp2p-request-response` does the closing for the sender, no need to do it in `Codec::send_request`). https://github.com/libp2p/rust-libp2p/blob/c224ffe01266e33101b609c10772fc116f7cf929/protocols/request-response/src/handler/protocol.rs#L172-L174
1. The receiver can read until the sender closes the stream or `REQUEST_SIZE_MAXIMUM` is reached. After sending the response, the receiver closes the stream. Again done by `libp2p-request-response`. https://github.com/libp2p/rust-libp2p/blob/c224ffe01266e33101b609c10772fc116f7cf929/protocols/request-response/src/handler/protocol.rs#L111-L114
2. The sender reads until the reading side is closed.

Does that make sense?

Reply:
> The receiver can read until the sender closes the stream or REQUEST_SIZE_MAXIMUM is reached.

This is the issue. If we don't know how much data we are going to read, we have to dynamically allocate memory:

1. Pre-allocating `REQUEST_SIZE_MAXIMUM` is definitely a no-go because a remote could force us to allocate a lot of memory but never send anything.
2. We can allocate in steps (which steps?) and stop if we receive `REQUEST_SIZE_MAXIMUM`.

Do you see any other option? (2) is a bit annoying to implement because `read_to_end` from `futures` unfortunately does not take a maximum size.

Reply:
1. As I understand, this problem is not only in `serde` codecs, so it looks reasonable to create a separate issue to fix this one. Ok?
2. Why is it wrong to write size in the message?

Reply:
> 2. Why is it wrong to write size in the message?

It is redundant :)
We already close the stream after writing the request / response.

Thus, there isn't a technical need for us to write the size.

> 1. As I understand, this problem is not only in `serde` codecs, so it looks reasonable to create a separate issue to fix this one. Ok?

We can't just change it later because that would be a breaking change on the network layer :(

Reply:
> Thanks for explaining.
> 
> How about mimicking what `Vec` already does, though with our additional requirement, i.e. not to exceed `REQUEST_SIZE_MAXIMUM`? As far as I can tell `Vec` simply grows exponentially. We can do the same here, right?
> 
> https://doc.rust-lang.org/src/alloc/raw_vec.rs.html#397-399

That is essentially `read_to_end` but with a max size right?

I think that is the way to go. Might even be worth asking upstream to always require a max size for security reasons.

Reply:
We're going to have our own `read_to_end` impl?

> Might even be worth asking upstream to always require a max size for security reasons.

Could you please give more detail?

Reply:
Looking through the issues on `futures-rs` I've found https://github.com/rust-lang/futures-rs/issues/2616.

That mentions a way of how we can easily implement this! We can simply read in steps of N bytes by using `take(N)` so no need to implement our own version of `read_to_end`! :)

Reply:
>> Why is it wrong to write size in the message?

> It is redundant :)
> We already close the stream after writing the request / response.

One moment for confidence.
Is this mean that we already got all the request/response data?
If it is, could we check the max size in the process of data getting?

Reply:
> That mentions a way of how we can easily implement this! We can simply read in steps of N bytes by using `take(N)` so no need to implement our own version of `read_to_end`! :)

Using `take` is a good idea. Thanks for the pointer. Just to make sure we are on the same page, I don't think we need to read "in steps". Instead one could do the below.

``` rust
stream.take(REQUEST_SIZE_MAXIMUM).read_to_end(buf).await;
```